wave_vision/include/wave/vision/tracker/tracker.hpp

```diff
@@ -0,0 +1,171 @@
+#include "wave/utils/utils.hpp"
+#include "wave/vision/utils.hpp"
+
+namespace wave {
+/** @addtogroup vision
+ *  @{ */
+
+class FeatureTrack {
+ public:
+    // Constructor
+    FeatureTrack() {}
+
+    // Initializer list constructor
+    FeatureTrack(size_t id,
+                 std::vector<LandmarkMeasurement<int>> measurements,
+                 size_t first_image,
+                 size_t last_image)
+        : id(id),
+          measurements(measurements),
+          first_image(first_image),
+          last_image(last_image) {}
+
+    // Destructor
+    ~FeatureTrack() = default;
+
+    /** Returns size of the measurement vector.
+     *
+     * @return size of the measurement vector
+     */
+    size_t size() const {
+        return this->measurements.size();
+    }
+
+    /** The id of the feature */
+    size_t id;
+
+    /** All measurements of the feature */
+    std::vector<LandmarkMeasurement<int>> measurements;
+
+    /** The first image the feature is seen in */
+    size_t first_image;
+
+    /** The last image the feature is seen in */
+    size_t last_image;
+};
+
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+class Tracker {
+ public:
+    /** Default constructor
+     *
+     * @param detector detector object (FAST, ORB, etc...)
+     * @param descriptor descriptor object (BRISK, ORB, etc...)
+     * @param matcher matcher object (BruteForceMatcher, FLANN)
+     */
+    Tracker(TDetector detector, TDescriptor descriptor, TMatcher matcher)
+        : detector(detector), descriptor(descriptor), matcher(matcher) {}
+
+    /** Destructor */
+    ~Tracker() = default;
+
+    /** Get the tracks of all features in the requested image from the sequence.
+     *
+     * @param img_num the number of the image to obtain tracks from
+     * @return tracks corresponding to all detected landmarks in the image.
+     */
+    std::vector<FeatureTrack> getTracks(const size_t &img_num) const;
+
+    /** Track features within an image (presumably the next in a sequence).
+     *
+     * @param image the image to add.
+     */
+    void addImage(const cv::Mat &image);
+
+    /** Draw tracks for the requested image.
+     *
+     * @param img_num the number of the image within the sequence
+     * @param image the image to draw the tracks on
+     * @return the image with the tracks illustrated as arrows.
+     */
+    cv::Mat drawTracks(const size_t &img_num, const cv::Mat &image) const;
+
+    /** Offline feature tracking, using list of images already loaded.
+     *
+     * @param image_sequence the sequence of images to analyze.
+     * @return the vector of FeatureTracks in each image.
+     */
+    std::vector<std::vector<FeatureTrack>> offlineTracker(
+      const std::vector<cv::Mat> &image_sequence);
+
+ private:
+    /** Generate a new ID for each newly detected feature.
+     *
+     * @return the assigned ID.
+     */
+    size_t generateFeatureID() const {
+        static size_t id = 0;
```

Comment:
I think that's fine. That's a good point, I wasn't aware of that. The actual ID assigned should not really matter as long as its unique...do you think it's okay as is?

PS thanks, but I stole this use of static from your MSCKF @leokoppel :smile: